components/service/glean/src/main/java/mozilla/components/service/glean/storages/GenericScalarStorageEngine.kt

```diff
@@ -166,6 +214,15 @@ internal abstract class GenericScalarStorageEngine<ScalarType> : StorageEngine {
         if (clearStore) {
             // We only allow clearing metrics with the "ping" lifetime.
             dataStores[Lifetime.Ping.ordinal].remove(storeName)
+
+            // Also remove from SharedPreferences
+            val editor = pingLifetimeStorage.edit()
+            pingLifetimeStorage.all.forEach {
+                if (it.key.startsWith("$storeName#")) {
+                    editor.remove(it.key)
+                }
+            }
+            editor.apply()
```

Comment:
If you iterated through the keys in `dataStores[Lifetime.Ping.ordinal]` and removed `$storeName#$key` from `pingLifetimeStorage`, this would be `O(n)` where `n` is the number of keys in the current store, rather than `O(m)` where `m` is the number of keys in all stores.  `m > n` should always hold true, and you have to do the hash lookups to remove the entries in the SharedPreferences either way.

Reply:
I like this approach, but I had to iterate through `dataStores[Lifetime.Ping.ordinal][storeName].keys` to get to the right level to accomplish this.